src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```diff
@@ -0,0 +1,192 @@
+package org.hydev.mcpm.client.database.fetcher;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.hc.client5.http.fluent.Request;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.HttpEntity;
+import org.hydev.mcpm.client.models.Database;
+import org.hydev.mcpm.utils.HashUtils;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * A database fetcher that has a local persistent cache along and a network fallback.
+ */
+public class LocalDatabaseFetcher implements DatabaseFetcher {
+    private final URI host;
+    private final Path cacheDirectory;
+
+    private Database localDatabase;
+
+    public static final String HASH_FILE_NAME = "db.hash";
+    public static final String DATABASE_FILE_NAME = "db";
+
+    public static final String USER_AGENT = "MCPM Client";
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     * The cache directory will default to .mcpm
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     */
+    public LocalDatabaseFetcher(URI host) {
+        this(host, Path.of(".mcpm/"));
```

Comment:
Hmm... should this be in a local folder? Or should it be in the user home directory like how I did in MirrorSelector?

For one part, I think it would be nice if we don't need to sync database again running the program in different locations, but for the other part, do you think home directory is an assumption we're making that some servers might not have?

Reply:
It doesn't matter. There's no restriction on where the cache directory needs to be. There's a constructor below this that allows you to specify where you want to put your cache directory. You can treat this constructor as a demo class if you'd like, I know this directory is supposed to be used for the scraper.